dask/dataframe/shuffle.py

```diff
@@ -376,52 +388,136 @@ def set_partition(
     return df4.map_partitions(M.sort_index)
 
 
-def shuffle(
+class ShuffleRegistry:
+    """A registry of methods for shuffling a dataframe."""
+
+    class Method(NamedTuple):
+        can_use: Callable[[Optional[str], Any, List[str]], bool]
+        method: Callable
+        priority: int
+
+    def __init__(self):
+        self.methods: Set[ShuffleRegistry.Method] = set()
+
+    @property
+    def known_priorities(self):
+        return {m.priority for m in self.methods}
+
+    def register(
+        self,
+        priority: int,
+        can_use: Callable[[Optional[str], Any, List[str]], bool],
+        method: Callable,
+    ):
+        """Register a new shuffle method
+
+        Parameters
+        ----------
+        priority
+            The priority of this method, higher priority methods are tried first
+        can_use
+            Function to check if this method is useable for the given
+            shuffle parameters. Receives the requested shuffle name
+            (e.g. "tasks", "disk"), the index to shuffle on, and a
+            list of dataframe column names. Should return True if this
+            method can shuffle the dataframe, and False otherwise.
+        method
+            Function to perform the shuffle, should have the same
+            signature as :func:`shuffle`.
+
+        Raises
+        ------
+        RuntimeError
+            If the requested priority is already in use for another method.
+        """
+        if priority in self.known_priorities:
+            raise RuntimeError(
+                "Registering multiple shuffle methods with same priority "
+                "is undefined behaviour"
+            )
+        self.methods.add(
+            ShuffleRegistry.Method(can_use=can_use, method=method, priority=priority)
+        )
+
+    def get(
+        self, shuffle_name: Optional[str], index: Any, df_columns: List[str]
+    ) -> Callable:
+        """Return a function to perform a shuffle
+
+        Parameters
+        ----------
+        shuffle_name
+            Requested shuffle type (e.g. "tasks", "disk")
+        index
+            The index to shuffle on
+        df_columns
+            List of column names in the dataframe being shuffled
+
+        Returns
+        -------
+        Function to perform the shuffle (with the same signature as
+        :func:`shuffle`).
+
+        Raises
+        ------
+        RuntimeError
+            If no matching shuffle method was found.
+        """
+        for method in sorted(self.methods, key=attrgetter("priority")):
+            if method.can_use(shuffle_name, index, df_columns):
+                return method.method
+        raise RuntimeError("Unreachable, didn't find a shuffle method")
+
+
+def _task_shuffle_no_partitions(
```

Comment:
Maybe the underlying shuffle method should just be responsible for deciding whether or not it should create/drop a "_partitions" column?

I get that the current `dask.dataframe` design tries to simplify things by (usually) creating a single `"_partitions"` column before calling down to `rearrange_by_column`. However, it seems like it would be a lot simpler if `dask.dataframe` just defined a clearly-documented function to generate this "_partitions" column (and required the underlying method to choose whether or not to use it).